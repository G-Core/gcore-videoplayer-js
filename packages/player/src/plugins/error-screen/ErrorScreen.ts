import { UICorePlugin, Events, template } from '@clappr/core'
import { trace } from '@gcorevideo/utils'

import { CLAPPR_VERSION } from '../../build.js'

import reloadIcon from '../../../assets/icons/old/reload.svg'
import templateHtml from '../../../assets/error-screen/error_screen.ejs'
import '../../../assets/error-screen/error_screen.scss'
import { PlaybackError } from '../../playback.types.js'

type ErrorScreenDesc = {
  title: string
  message: string
  code: string
  icon?: string
}

/**
 * Configuration options for the {@link ErrorScreen | error screen} plugin.
 * @beta
 */
export type ErrorScreenPluginSettings = {
  /**
   * Whether to hide the reload button.
   */
  noReload?: boolean
}

const T = 'plugins.error_screen'

/**
 * `PLUGIN` that displays errors nicely in the overlay on top of the player.
 * @beta
 */
export class ErrorScreen extends UICorePlugin {
  private err: ErrorScreenDesc | null = null

  /**
   * @internal
   */
  get name() {
    return 'error_screen'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  private static readonly template = template(templateHtml)

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'player-error-screen',
      'data-error-screen': '',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenTo(this.core, Events.ERROR, this.onError)
    this.listenTo(
      this.core,
      Events.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onActiveContainerChanged,
    )
  }

  private onPlay() {
    trace(`${T} onPlay`)
    this.unmount()
  }

  private unmount() {
    trace(`${T} unmount`)
    this.err = null
    this.$el.remove()
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click .player-error-screen__reload': 'reload',
    }
  }

  private reload() {
    setTimeout(() => {
      this.core.configure({
        reloading: true,
        source: this.core.options.source,
        sources: this.core.options.sources,
      })
    }, 0)
  }

  private onActiveContainerChanged() {
    trace(`${T} onActiveContainerChanged`, {
      reloading: this.core.options.reloading,
    })
    this.err = null
    this.listenTo(
      this.core.activeContainer.playback,
      Events.PLAYBACK_PLAY,
      this.onPlay,
    )
    if (this.core.options.reloading) {
      setTimeout(() => {
        this.core.options.reloading = false
        this.unmount()
        this.core.activeContainer.play({
          reloading: true,
        })
      }, 0)
    }
  }

  private onError(err: PlaybackError) {
    trace(`${T} onError`, { err })
    if (err.UI) {
      if (this.err) {
        this.unmount()
      }
      this.err = {
        title: err.UI.title,
        message: err.UI.message,
        code: err.code,
        icon: err.UI.icon,
      }
      this.render()
    }
  }

  /**
   * @internal
   */
  override render() {
    if (!this.err) {
      return this
    }
    this.$el.html(
      ErrorScreen.template({
        ...this.err,
        reloadIcon: this.options.errorScreen?.noReload ? null : reloadIcon,
      }),
    )

    // TODO append to container instead of core?
    if (!this.el.parentElement) {
      this.core.$el.append(this.el)
    }

    return this
  }
}
