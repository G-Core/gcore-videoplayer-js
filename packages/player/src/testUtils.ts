import Events from 'eventemitter3'
import { vi } from 'vitest'
/**
 * @internal
 * @deprecated
 * TODO use createMockPlayback() instead
 */
export class _MockPlayback extends Events {
  constructor(
    protected options: any,
    readonly i18n: any,
    protected playerError?: any,
  ) {
    super()
  }

  get name() {
    return 'mock'
  }

  consent() {}

  play() {}

  pause() {}

  stop() {}

  destroy() {}

  seek() {}

  seekPercentage() {}

  getDuration() {
    return 100
  }

  enterPiP() {}

  exitPiP() {}

  getPlaybackType() {
    return 'live'
  }

  getStartTimeOffset() {
    return 0
  }

  getCurrentTime() {
    return 0
  }

  isHighDefinitionInUse() {
    return false
  }

  mute() {}

  unmute() {}

  volume() {}

  configure() {}

  attemptAutoPlay() {
    return true
  }

  canAutoPlay() {
    return true
  }

  onResize() {
    return true
  }

  trigger(event: string, ...args: any[]) {
    this.emit(event, ...args)
  }
}

export function createMockCore(options: Record<string, unknown> = {}, container: any = createMockContainer()) {
  const el = document.createElement('div')
  return Object.assign(new Events(), {
    el,
    $el: {
      [0]: el,
      append: vi.fn(),
    },
    activePlayback: container.playback,
    activeContainer: container,
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    levels: [],
    consent() {},
    play() {},
    pause() {},
    stop() {},
    destroy() {},
    seek() {},
    seekPercentage() {},
    getDuration() {
      return 100
    },
    enterPiP() {},
    exitPiP() {},
    getPlaybackType() {
      return 'live'
    },
    getStartTimeOffset() {
      return 0
    },
    getCurrentTime() {
      return 0
    },
    isHighDefinitionInUse() {
      return false
    },
    mute() {},
    unmute() {},
    volume() {},
    configure() {},
    attemptAutoPlay() {
      return true
    },
    canAutoPlay() {
      return true
    },
    onResize() {
      return true
    },
    trigger(event: string, ...args: any[]) {
      emitter.emit(event, ...args)
    },
  })
}

export function createMockContainer(playback: any = createMockPlayback()) {
  const el = document.createElement('div')
  return Object.assign(new Events(), {
    $el: {
      html: vi.fn(),
      [0]: el,
    },
    el,
    getPlugin: vi.fn(),
    playback,
  })
}
