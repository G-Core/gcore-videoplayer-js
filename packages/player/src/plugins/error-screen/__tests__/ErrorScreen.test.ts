import { createMockCore } from '../../../testUtils'
import { ErrorScreen } from '../ErrorScreen'
import { beforeEach, describe, expect, it, vi } from 'vitest'

describe('ErrorScreen', () => {
  let core: any
  let container: any
  let playback: any
  let errorScreen: ErrorScreen
  beforeEach(() => {
    core = createMockCore()
    container = core.activeContainer
    playback = container.playback
    core.activeContainer = container
  })
  describe('on error', () => {
    describe.each([
      [
        'targeted at UI',
        {
          code: "foo",
          UI: {
            title: 'My error',
            message: 'Lorem ipsum dolor sit amet',
            icon: '<svg>...</svg>'
          },
        },
        true,
      ],
      [
        'other',
        {
          code: 'MEDIA_SOURCE_ACCESS_DENIED',
          title: 'The server refused to serve the protected media',
          message: 'Wrong credentials',
        },
        false,
      ],
    ])("%s", (_, err, shouldRender) => {
      beforeEach(() => {
        errorScreen = new ErrorScreen(core)
        core.emit('core:ready')
        core.emit('core:active:container:changed')
        core.emit('error', err)
      })
      if (shouldRender) {
        it('should render', () => {
          expect(errorScreen.el.innerHTML).toBeTruthy()
          expect(errorScreen.el.innerHTML).toMatchSnapshot()
        })
      } else {
        it('should not render', () => {
          expect(errorScreen.el.innerHTML).toBeFalsy()
        })
      }
    })
    describe('reload button', () => {
      describe('basically', () => {
        beforeEach(() => {
          core.options.source = 'https://222/master.mpd'
          errorScreen = new ErrorScreen(core)
          core.emit('core:ready')
          core.emit('core:active:container:changed')
          core.emit('error', {
            code: 'foo',
            UI: {
              title: 'My error',
              message: 'Lorem ipsum dolor sit amet',
              icon: '<svg>...</svg>'
            },
          })
          core.configure.mockClear()
        })
        describe('when clicked', () => {
          beforeEach(async () => {
            (errorScreen.el.querySelector('.player-error-screen__reload') as HTMLElement)?.click()
            return new Promise(resolve => setTimeout(resolve, 0))
          })
          it('should reload the player', () => {
            expect(core.configure).toHaveBeenCalledWith(expect.objectContaining({
              reloading: true,
              source: 'https://222/master.mpd'
            }))
          })
        })
      })
      describe('when disabled', () => {
        beforeEach(() => {
          core.options.errorScreen = {
            noReload: true
          }
          errorScreen = new ErrorScreen(core)
          core.emit('core:ready')
          core.emit('core:active:container:changed')
          core.emit('error', {
            code: 'foo',
            UI: {
              title: 'My error',
              message: 'Lorem ipsum dolor sit amet',
              icon: '<svg>...</svg>'
            },
          })
        })
        it('should not render the reload button', () => {
          expect(errorScreen.el.querySelector('.player-error-screen__reload')).toBeNull()
        })
      })
    })
  })
})


