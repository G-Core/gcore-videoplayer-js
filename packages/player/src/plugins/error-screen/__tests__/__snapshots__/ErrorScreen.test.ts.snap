// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorScreen > on error > targeted at UI > should render 1`] = `
"<div class="player-error-screen__content" data-error-screen="">
    
        <div class="player-error-screen__icon" data-error-screen=""><svg>...</svg></div>
    
    <div class="player-error-screen__title" data-error-screen="">My error</div>
    
        <div class="player-error-screen__message" data-error-screen="">Lorem ipsum dolor sit amet</div>
    
    
        <div class="player-error-screen__code" data-error-screen="">Error code: foo</div>
    
    
        <div class="player-error-screen__reload" data-error-screen="">/assets/icons/old/reload.svg</div>
    
</div>
"
`;
